src/wasm/wasm-binary.cpp

```diff
@@ -2109,13 +2104,13 @@ Signature WasmBinaryBuilder::getSignatureByTypeIndex(Index index) {
   if (!heapType.isSignature()) {
     throwError("invalid signature type " + heapType.toString());
   }
-  return heapType.getSignature();
+  return heapType;
 }
 
 void WasmBinaryBuilder::readFunctions() {
   BYN_TRACE("== readFunctions\n");
   size_t total = getU32LEB();
-  if (total != functionSignatures.size()) {
+  if (total != functionSignatures.size() - functionImports.size()) {
```

Comment:
this change seems very surprising, as I thought the PR is an NFC change that just stores heap types instead of signatures, but keeps indexing all the same..?

Reply:
The imported `Function`s don't contain their HeapTypes, so we need to change `functionSignatures` (soon to be renamed `functionTypes`) to include the HeapTypes of the imported functions as well to avoid having to resynthesize them later.